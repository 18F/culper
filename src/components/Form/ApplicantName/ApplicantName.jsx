import React from 'react'
import ValidationElement from '../validationElement'
import Text from '../Text'
import Dropdown from '../Dropdown'
import { api } from '../../../services/api'

export default class Address extends ValidationElement {
  constructor (props) {
    super(props)

    this.state = {
      name: props.name,
      label: props.label,
      first: props.first,
      last: props.last,
      middle: props.middle,
      suffix: props.suffix,
      suffixOther: props.suffixOther,
      focus: props.focus || false,
      error: props.error || false,
      valid: props.valid || false
    }
  }

  /**
   * Handle the change event.
   */
  handleChange (event) {
    let part = this.extractPart(event.target.id)
    let value = event.target.value
    let updated = null

    switch (part) {
      case 'first':
        updated = { first: value }
        break
      case 'last':
        updated = { last: value }
        break
      case 'middle':
        updated = { middle: value }
        break
      case 'suffix':
        updated = { suffix: value }
        break
      case 'suffixOther':
        updated = { suffixOther: value }
        break
    }

    this.setState(updated, () => {
      super.handleChange(event)
    })
  }

  /**
   * Handle the validation event.
   */
  handleValidation (event, status) {
    this.setState({error: status === false, valid: status === true}, () => {
      if (this.state.error === false || this.state.valid === true) {
        super.handleValidation(event, status)
        return
      }

      api
        .validateApplicantName({
          Last: this.state.last,
          First: this.state.first,
          Middle: this.state.middle,
          Suffix: this.state.suffix,
          SuffixOther: this.state.suffixOther
        })
        .then((response) => {
          // TODO: Display and assign the errors as necessary
          if (response.Errors) {
          }

          if (response.Suggestions) {
          }
        })
        .then(() => {
          super.handleValidation(event, status)
        })
    })
  }

  /**
   * Generated name for the part of the address elements.
   */
  partName (part) {
    return '' + this.state.name + '-' + part
  }

  /**
   * Returns the part name from the pull generated identifier.
   */
  extractPart (id) {
    return id.split('-').pop()
  }

  /**
   * Toggles visibility class for the extended suffix element.
   */
  suffixOtherClass () {
    return this.state.suffix === 'Other'
      ? ''
      : 'hidden'
  }

  render () {
    return (
      <div>
        <Text name={this.partName('last')}
              label="Last name"
              maxlength="100"
              pattern="^[a-zA-Z\-\.' ]*$"
              help="The last name is required, cannot exceed 100 characters, and we only support letters, hyphens (-), periods (.), apostrophes ('), and spaces."
              value={this.state.last}
              onChange={this.handleChange}
              onValidate={this.handleValidation}
              onFocus={this.props.onFocus}
              onBlur={this.props.onBlur}
              />
        <Text name={this.partName('first')}
              label="First name"
              maxlength="100"
              help="The first name (or initial) is optional but cannot exceed 100 characters"
              value={this.state.first}
              onChange={this.handleChange}
              onValidate={this.handleValidation}
              onFocus={this.props.onFocus}
              onBlur={this.props.onBlur}
              />
        <Text name={this.partName('middle')}
              label="Middle name or initial"
              minlength="0"
              maxlength="100"
              help="The middle name (or initial) is optional but cannot exceed 100 characters"
              value={this.state.middle}
              onChange={this.handleChange}
              onValidate={this.handleValidation}
              onFocus={this.props.onFocus}
              onBlur={this.props.onBlur}
              />
        <Dropdown name={this.partName('suffix')}
                  label="Suffix"
                  value={this.state.suffix}
                  onChange={this.handleChange}
                  onValidate={this.handleValidation}
                  onFocus={this.props.onFocus}
                  onBlur={this.props.onBlur}
                  >
          <option value=""></option>
          <option value="Jr">Jr</option>
          <option value="Sr">Sr</option>
          <option value="I">I</option>
          <option value="II">II</option>
          <option value="III">III</option>
          <option value="IV">IV</option>
          <option value="V">V</option>
          <option value="VI">VI</option>
          <option value="VII">VII</option>
          <option value="VIII">VIII</option>
          <option value="IX">IX</option>
          <option value="X">X</option>
          <option value="Other">Other</option>
        </Dropdown>
        <div className={this.suffixOtherClass()}>
          <Text name={this.partName('suffixOther')}
                label="Other"
                maxlength="100"
                value={this.state.suffixOther}
                onChange={this.handleChange}
                onValidate={this.handleValidation}
                onFocus={this.props.onFocus}
                onBlur={this.props.onBlur}
                />
        </div>
      </div>
    )
  }
}
