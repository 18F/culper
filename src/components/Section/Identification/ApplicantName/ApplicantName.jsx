import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import validate from '../../../../validators'
import SubsectionElement from '../../SubsectionElement'
import { Name, Field } from '../../../Form'

export default class ApplicantName extends SubsectionElement {
  constructor (props) {
    super(props)
    this.update = this.update.bind(this)
    this.updateName = this.updateName.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      Name: this.props.Name,
      ...queue
    })
  }

  updateName (values) {
    this.update({
      Name: values
    })
  }

  render () {
    const klass = `applicant-name ${this.props.className || ''}`.trim()

    return (
      <div className={klass}>
        <Field title={i18n.t('identification.name.title')}
               titleSize="h2"
               optional={true}
               scrollIntoView={this.props.scrollIntoView}>
          <Name name="name"
                {...this.props.Name}
                dispatch={this.props.dispatch}
                onUpdate={this.updateName}
                onError={this.handleError}
                required={this.props.required}
                scrollIntoView={this.props.scrollIntoView}
                />
        </Field>
      </div>
    )
  }
}

ApplicantName.defaultProps = {
  Name: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'name',
  dispatch: () => {},
  required: false,
  validator: (data) => {
    return validate(schema('identification.name', data))
  }
}

ApplicantName.errors = []
