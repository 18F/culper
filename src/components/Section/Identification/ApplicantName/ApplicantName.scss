.name-label {
	margin-bottom: 2rem;
}