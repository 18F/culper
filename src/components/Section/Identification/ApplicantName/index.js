import ApplicantName from './ApplicantName'
export default ApplicantName
