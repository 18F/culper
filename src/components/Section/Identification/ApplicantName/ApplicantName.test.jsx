import React from 'react'
import { shallow, mount } from 'enzyme'
import { ApplicantName } from './ApplicantName'

describe('The ApplicantName component', () => {
  it('renders without errors', () => {
    const component = shallow(<ApplicantName />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it('displays no error by default', () => {
    const component = mount(<ApplicantName />)

    component.find('.first input').simulate('change')
    expect(component.find('.usa-input-error-label').length).toEqual(0)
  })

  it('implements an onUpdate handler', () => {
    let updates = 0
    const testProps = {
      onUpdate: () => {
        updates += 1
      },
    }

    const component = mount(<ApplicantName {...testProps} />)
    component.find('.first input').simulate('change')
    expect(updates).toBe(1)
  })
})
