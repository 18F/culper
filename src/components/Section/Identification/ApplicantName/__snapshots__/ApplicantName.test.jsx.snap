// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ApplicantName component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content applicant-name"
    data-section="identification"
    data-subsection="name"
  >
    <h1
      className="section-header"
    >
      Full name
    </h1>
    <Field
      adjustFor=""
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help=""
      helpActive={false}
      optional={true}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Provide your full name"
      titleSize="h4"
      validate={true}
    >
      <Name
        error={false}
        errorCodes={Array []}
        first=""
        firstInitialOnly={false}
        focus={false}
        hideMiddleName={false}
        last=""
        middle=""
        middleInitialOnly={false}
        name="name"
        noMiddleName={false}
        onError={[Function]}
        onUpdate={[Function]}
        prefix="name"
        required={false}
        suffix=""
        suffixOther=""
        valid={false}
      />
    </Field>
  </div>,
  "nodes": Array [
    <div
      className="section-content applicant-name"
      data-section="identification"
      data-subsection="name"
    >
      <h1
        className="section-header"
      >
        Full name
      </h1>
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help=""
        helpActive={false}
        optional={true}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Provide your full name"
        titleSize="h4"
        validate={true}
      >
        <Name
          error={false}
          errorCodes={Array []}
          first=""
          firstInitialOnly={false}
          focus={false}
          hideMiddleName={false}
          last=""
          middle=""
          middleInitialOnly={false}
          name="name"
          noMiddleName={false}
          onError={[Function]}
          onUpdate={[Function]}
          prefix="name"
          required={false}
          suffix=""
          suffixOther=""
          valid={false}
        />
      </Field>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ApplicantName
        Name={Object {}}
        dispatch={[Function]}
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        validator={[Function]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ApplicantName {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "Name": Object {},
          "dispatch": [Function],
          "onError": [Function],
          "onUpdate": [Function],
          "required": false,
          "validator": [Function],
        },
        "refs": Object {},
        "section": "identification",
        "state": null,
        "store": "Identification",
        "storeKey": "ApplicantName",
        "subsection": "name",
        "update": [Function],
        "updateName": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content applicant-name"
          data-section="identification"
          data-subsection="name"
        >
          <h1
            className="section-header"
          >
            Full name
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={true}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your full name"
            titleSize="h4"
            validate={true}
          >
            <Name
              error={false}
              errorCodes={Array []}
              first=""
              firstInitialOnly={false}
              focus={false}
              hideMiddleName={false}
              last=""
              middle=""
              middleInitialOnly={false}
              name="name"
              noMiddleName={false}
              onError={[Function]}
              onUpdate={[Function]}
              prefix="name"
              required={false}
              suffix=""
              suffixOther=""
              valid={false}
            />
          </Field>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="section-content applicant-name"
          data-section="identification"
          data-subsection="name"
        >
          <h1
            className="section-header"
          >
            Full name
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={true}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your full name"
            titleSize="h4"
            validate={true}
          >
            <Name
              error={false}
              errorCodes={Array []}
              first=""
              firstInitialOnly={false}
              focus={false}
              hideMiddleName={false}
              last=""
              middle=""
              middleInitialOnly={false}
              name="name"
              noMiddleName={false}
              onError={[Function]}
              onUpdate={[Function]}
              prefix="name"
              required={false}
              suffix=""
              suffixOther=""
              valid={false}
            />
          </Field>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ApplicantName
    Name={Object {}}
    dispatch={[Function]}
    onError={[Function]}
    onUpdate={[Function]}
    required={false}
    validator={[Function]}
  />,
}
`;
